app/views/tag_substitutions/new.html.erb

```diff
@@ -37,15 +37,28 @@
         </tr>
       </thead>
       <tbody>
-        <% @tag_substitution.substitutions.each do |substitution| %>
+        <% @tag_substitution.substitutions.each_with_index do |substitution, index| %>
           <%= form.fields_for :substitutions, substitution, index: nil do |sub| %>
             <tr>
-              <td><%= sub.hidden_field :sample_id %><%= substitution.sample_id %>: <%= substitution.sample_friendly_name %></td>
-              <td><%= sub.hidden_field :library_id %><%= substitution.library_id %></td>
-              <td><%= sub.hidden_field :original_tag_id %><%= tag_name(substitution.original_tag_id) %></td>
-              <td><%= sub.select :substitute_tag_id, tag_options_for(substitution.substitute_tag_id), {}, class: 'select2 tag-list' %></td>
-              <td><%= sub.hidden_field :original_tag2_id %><%= tag_name(substitution.original_tag2_id) %></td>
-              <td><%= sub.select :substitute_tag2_id, tag_options_for(substitution.substitute_tag2_id), {}, class: 'select2 tag-list' %></td>
+              <td>
+                <%= sub.hidden_field :sample_id, id: "ts_s_id_#{index}" %>
+                <%= substitution.sample_id %>: <%= substitution.sample_friendly_name %>
+              </td>
+              <td><%= sub.hidden_field :library_id, id: "ts_l_id_#{index}"  %><%= substitution.library_id %></td>
+              <td>
+                <%= sub.hidden_field :original_tag_id, id: "ts_ot_id_#{index}" %>
+                <%= tag_name(substitution.original_tag_id) %>
+              </td>
+              <td><%= sub.select :substitute_tag_id, tag_options_for(substitution.substitute_tag_id),
+                                 {},
+                                 class: 'tag-list', id: "ts_st_id_#{index}"  %></td>
+              <td>
+                <%= sub.hidden_field :original_tag2_id, id: "ts_ot2_id_#{index}" %>
+                <%= tag_name(substitution.original_tag2_id) %>
+              </td>
+              <td><%= sub.select :substitute_tag2_id, tag_options_for(substitution.substitute_tag2_id),
+                                 {},
+                                 class: 'tag-list' , id: "ts_st2_id_#{index}" %></td>
```

Comment:
Why are you adding the ids? Doesn't look like they are used?

Reply:
Otherwise the rails form-tags will generate their own ids, and unfortunately they generate the same id for each set of substitution fields. Not only does this result in lots of warnings in the developer console, but also stops the select2 extension from properly binding to the dropdowns. (Select2 provides a search box inside the select field, which makes it much easier to find the relevant tag)

So really just adding an id to ensure they are unique, and make sure the HTML remains valid.